src/WinRT.Runtime/IWinRTObject.net5.cs

```diff
@@ -1,188 +1,270 @@
-                {
-                    ExceptionHelpers.ThrowExceptionForHR(hr);
-                }
-                return false;
-            }
-
-            if (typeof(System.Collections.IEnumerable).IsAssignableFrom(type))
-            {
-                RuntimeTypeHandle projectIEnum = typeof(System.Collections.IEnumerable).TypeHandle;
-                AdditionalTypeData.GetOrAdd(projectIEnum, (_) => new ABI.System.Collections.IEnumerable.AdaptiveFromAbiHelper(type, this));
-            }
-
-            var vftblType = helperType.FindVftblType();
-            using (objRef)
-            {
-                if (vftblType is null)
-                {
-                    var qiObjRef = objRef.As<IUnknownVftbl>(GuidGenerator.GetIID(helperType));
-                    if (!QueryInterfaceCache.TryAdd(interfaceType, qiObjRef))
-                    {
-                        qiObjRef.Dispose();
-                    }
-                    return true;
-                }
-                IObjectReference typedObjRef = (IObjectReference)typeof(IObjectReference).GetMethod("As", Type.EmptyTypes).MakeGenericMethod(vftblType).Invoke(objRef, null);
-                if (!QueryInterfaceCache.TryAdd(interfaceType, typedObjRef))
-                {
-                    typedObjRef.Dispose();
-                }
-                return true;
-            }
-        }
-
-        RuntimeTypeHandle IDynamicInterfaceCastable.GetInterfaceImplementation(RuntimeTypeHandle interfaceType)
-        {
-            var type = Type.GetTypeFromHandle(interfaceType);
-            var helperType = type.GetHelperType();
-            if (helperType.IsInterface)
-                return helperType.TypeHandle;
-            return default;
-        }
-
-        IObjectReference NativeObject { get; }
-        bool HasUnwrappableNativeObject { get; }
-
-        protected ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> QueryInterfaceCache { get; }
-
-        IObjectReference GetObjectReferenceForType(RuntimeTypeHandle type)
-        {
-            return GetObjectReferenceForTypeFallback(type);
-        }
-
-        IObjectReference GetObjectReferenceForTypeFallback(RuntimeTypeHandle type)
-        {
-            if (IsInterfaceImplemented(type, true))
-            {
-                return QueryInterfaceCache[type];
-            }
-            throw new Exception("Interface " + Type.GetTypeFromHandle(type) +" is not implemented.");
-        }
-
-        ConcurrentDictionary<RuntimeTypeHandle, object> AdditionalTypeData { get; }
-
-        object GetOrCreateTypeHelperData(RuntimeTypeHandle type, Func<object> helperDataFactory)
-        {
-            return AdditionalTypeData.GetOrAdd(type, (type) => helperDataFactory());
-        }
-    }
-}
\ No newline at end of file
+#if EMBED
+    internal
+#else
+    public
+#endif
+    interface IWinRTObject : IDynamicInterfaceCastable
+#if NET8_0_OR_GREATER
+        , IUnmanagedVirtualMethodTableProvider

```

Comment:
I will need to check if adding an interface here is considered an API compat breaking change or if it is fine if we also implement it.

Reply:
Theoretically, there could be a diamond problem if two interfaces were to both implement `IWinRTObject` and then a class were to implement both of those interfaces. Other than that case (which shouldn't happen given the expectation of where this interface is implemented), this should be fine.

Alternatively, `IWinRTObject`'s `IDynamicInterfaceCastable` implementation could also provide an implementation for this interface lazily.